[14.0][IMP] stock_reception_screen_measuring_device: measure smaller packages

mmequignon opened the pull request:
When goods are received, triggers measurements also for smaller packagings that weren't ordered or received.

This behavior is not triggered when received packagings are not the same as ordered packagings.

mmequignon committed:
stock_reception_screen_measuring_device: measure smaller packages

When goods are received, triggers measurements also for smaller
packagings that weren't ordered or received.

This behavior is not triggered when received packagings are not the same
as ordered packagings.

sebalix commented on stock_reception_screen_measuring_device/migrations/14.0.2.0.0/pre-migration.py:
```diff
@@ -0,0 +1,29 @@
+
+
+def create_and_populate_new_fields(cr):
+    cr.execute(
+        """
+        ALTER TABLE stock_reception_screen
+        ADD COLUMN IF NOT EXISTS smaller_package_has_missing_dimensions BOOLEAN;
+        """
+    )
```
Not blocking, but there are `odoo.tools.sql.create_column` helper for that.

simahawk commented on stock_reception_screen_measuring_device/models/stock_reception_screen.py:
```diff
@@ -30,13 +36,18 @@ class StockReceptionScreen(models.Model):
-                record.product_packaging_id
-                and record.product_packaging_id.measuring_device_id
-            )
+            # TODO
```
TODO what?

simahawk reviewed (approved):
LG

grindtildeath commented on stock_reception_screen_measuring_device/migrations/14.0.2.0.0/post-migration.py:
```diff
@@ -0,0 +1,25 @@
+    )
+    records_to_update._compute_smaller_package_has_missing_dimensions()
+
+
+def migrate(cr, version):
+    if not version:
+        return
+    env = api.Environment(cr, SUPERUSER_ID, {})
```
You can use `@openupgrade.migrate()` as decorator to get an environment as parameter instead of `cr` (not blocking)

grindtildeath commented on stock_reception_screen_measuring_device/models/stock_reception_screen.py:
```diff
@@ -30,13 +36,17 @@ class StockReceptionScreen(models.Model):
-                and record.product_packaging_id.measuring_device_id
-            )
+            all_product_packagings = record.current_move_product_id.packaging_ids
+            record.scan_requested = False
+            for packaging in all_product_packagings:
+                if packaging.measuring_device_id:
+                    record.scan_requested = True
+                    break
```
```suggestion
            record.scan_requested = any(packaging.measuring_device_id for packaging in all_product_packagings)
```

mmequignon replied:
Even if this is nicer to read, I think the for loop implementation is faster, because it does not evaluate the value every   record.

grindtildeath commented on stock_reception_screen_measuring_device/models/stock_reception_screen.py:
```diff
@@ -71,29 +95,62 @@ def _compute_package_has_missing_dimensions(self):
+    def _get_smaller_package_without_dimensions(self):
+        self.ensure_one()
+        domain = self._get_smaller_package_without_dimensions_domain()
+        return self.env["product.packaging"].search(domain, order="qty desc", limit=1)
```
Looks weird, if you want the "smaller" package, shouldn't you order by `qty asc` with the limit set to 1 ?

simahawk replied:
@mmequignon ping

mmequignon replied:
no, we measure from the bigger to the smallest
bix box -> medium box -> small box

mmequignon replied:
We want the bigger package among the smaller packages missing measurements.

grindtildeath reviewed (approved):
Looks good, small remarks

sebalix commented on stock_reception_screen_measuring_device/models/stock_reception_screen.py:
```diff
@@ -30,13 +36,17 @@ class StockReceptionScreen(models.Model):
 
-    @api.depends("product_packaging_id", "product_packaging_id.measuring_device_id")
+    @api.depends(
+        "current_move_product_id.packaging_ids.measuring_device_id",
```
We could trigger a lot of computation with such depends knowing that `scan_requested` field is `store=True`, especially each time we call `_measuring_device_assign`/`_measuring_device_release` methods.

IMO this `scan_requested` field could be set `store=False` as it's only used to display a button on the view. So this field will be computed only when needed, avoiding recomputation of its value on all `stock.reception.screen` records.

simahawk replied:
@mmequignon this sounds like a good point

simahawk replied:
fixed

simahawk committed:
stock_reception_screen_measuring_device: fix scan_requested stored

There's no need to store this flag as it's used only for UI purpose on the rec screen.
This way useless computation to update the field are avoided.

sebalix commented:
/ocabot merge nobump
